arch/riscv/kvm/vcpu_sbi_covh.c

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: GPL-2.0
+/*
+ * Copyright (c) 2024 IBM.
+ *
+ * Authors:
+ *     Wojciech Ozga <woz@zurich.ibm.com>
+ */
+
+#include <linux/errno.h>
+#include <linux/err.h>
+#include <linux/kvm_host.h>
+#include <linux/list.h>
+#include <linux/mm.h>
+#include <linux/spinlock.h>
+#include <asm/csr.h>
+#include <asm/sbi.h>
+#include <asm/kvm_vcpu_sbi.h>
+#include <asm/kvm_cove.h>
+#include <asm/kvm_cove_sbi.h>
+#include <asm/kvm_nacl.h>
+#include <linux/rbtree.h>
+#include <linux/pgtable.h>
+
+static int preload_pages(struct kvm_vcpu *vcpu) {
+	struct kvm_memory_slot *memslot;
+	unsigned long hva, fault_addr, page;
+	bool writable;
+
+	/* We must preload all initial pages so that during the promotion the TSM can
+	 * copy these pages to confidential memory and measure them. 
+	 */
+	memslot = search_memslots(kvm_memslots(vcpu->kvm), kernel_map.phys_addr, true);
+	if (memslot) {
+		for (page = 0; page < memslot->npages; page++) {
+			fault_addr = gfn_to_gpa(memslot->base_gfn) + page * PAGE_SIZE;
+			hva = gfn_to_hva_memslot_prot(memslot, gpa_to_gfn(fault_addr), &writable);
+			if (!kvm_is_error_hva(hva)) {
+				kvm_riscv_gstage_map(vcpu, memslot, fault_addr, hva, writable);
+			}
+		}
+	}
+
+	return 0;
+}
+
+static int kvm_riscv_cove_promote_to_tvm(struct kvm_vcpu *vcpu, 
+					unsigned long fdt_address, unsigned long tap_addr) {
+	int rc, bkt;
+	struct kvm_memory_slot *memslot;
+	preload_pages(vcpu);
+
+	rc = kvm_riscv_cove_vm_single_step_init(vcpu, fdt_address, tap_addr);
+	if (rc) {
+		goto done;
+	}
+
+	kvm_for_each_memslot(memslot, bkt, kvm_memslots(vcpu->kvm)) {
+		if (memslot->base_gfn == kernel_map.phys_addr)
+			kvm_arch_flush_shadow_memslot(vcpu->kvm, memslot);
+	}
+
```

Comment:
Once TSM successfully promotes a VM to TVM, we can safely free all VM's memory (it has been copied to confidential memory by TSM).